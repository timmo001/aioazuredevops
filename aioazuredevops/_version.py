"""Version helper for aioazuredevops.

Expose a simple string value for the current package version,
to remove the dependency on incremental and its auto-generated file.
"""

from __future__ import annotations

try:
    # Prefer to derive version from setup.py via importlib.metadata when installed
    from importlib.metadata import version as _dist_version

    __version__ = _dist_version("aioazuredevops")
except Exception:  # pragma: no cover - fallback to static string if not installed
    # Keep in sync with setup.py
    __version__ = "2.2.2.dev0"

__all__ = ["__version__"]
