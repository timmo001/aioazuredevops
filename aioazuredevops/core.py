"""
DevOps Core
https://docs.microsoft.com/en-gb/rest/api/azure/devops/core/?view=azure-devops-rest-6.0
"""
from datetime import datetime


class DevOpsLinks:
    """Links."""

    def __init__(self, linkSelf: str = None, linkCollection: str = None, linkWeb: str = None):
        """Initialize"""
        self.self = linkSelf
        self.collection = linkCollection
        self.web = linkWeb


class DevOpsTeam:
    """DevOps Team."""

    def __init__(self, id: str, name: str, url: str = None):
        """Initialize"""
        self.id = id
        self.name = name
        self.url = url


class DevOpsProject:
    """A DevOps Project"""

    def __init__(
        self,
        id: str,
        name: str,
        description: str = None,
        url: str = None,
        state: str = None,
        revision: int = None,
        visibility: str = None,
        last_updated: datetime = None,
        default_team: DevOpsTeam = None,
        links: DevOpsLinks = None,
    ):
        """Initialize"""
        self.id = id
        self.name = name
        self.description = description
        self.url = url
        self.state = state
        self.revision = revision
        self.visibility = visibility
        self.last_updated = last_updated
        self.default_team = default_team
        self.links = links
