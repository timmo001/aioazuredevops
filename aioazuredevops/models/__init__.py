"""Models."""

from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")


@dataclass
class ListResult[T]:
    """List result."""

    count: int
    value: list[T]
