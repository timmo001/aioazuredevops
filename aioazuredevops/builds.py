"""
DevOps Builds
https://docs.microsoft.com/en-gb/rest/api/azure/devops/build/builds?view=azure-devops-rest-6.0
"""
from aioazuredevops.core import DevOpsProject


class DevOpsBuildLinks:
    """DevOps build links."""

    def __init__(
        self,
        l_self: str = None,
        web: str = None,
        source_version_display_uri: str = None,
        timeline: str = None,
        badge: str = None,
    ):
        """Initialize"""
        self.self = l_self
        self.web = web
        self.source_version_display_uri = source_version_display_uri
        self.timeline = timeline
        self.badge = badge


class DevOpsBuildDefinition:
    """DevOps build definition."""

    def __init__(
        self,
        id: int,
        name: str,
        url: str = None,
        path: str = None,
        type: str = None,
        queue_status: str = None,
        revision: int = None,
    ):
        self.id = id
        self.name = name
        self.url = url
        self.path = path
        self.type = type
        self.queue_status = queue_status
        self.revision = revision


class DevOpsBuild:
    """DevOps build."""

    def __init__(
        self,
        id: int,
        build_number: str = None,
        status: str = None,
        result: str = None,
        source_branch: str = None,
        source_version: str = None,
        priority: str = None,
        reason: str = None,
        queue_time: str = None,
        start_time: str = None,
        finish_time: str = None,
        definition: DevOpsBuildDefinition = None,
        project: DevOpsProject = None,
        links: DevOpsBuildLinks = None,
    ):
        """Initialize"""
        self.id = id
        self.links = links
        self.build_number = build_number
        self.status = status
        self.result = result
        self.source_branch = source_branch
        self.source_version = source_version
        self.priority = priority
        self.reason = reason
        self.queue_time = queue_time
        self.start_time = start_time
        self.finish_time = finish_time
        self.definition = definition
        self.project = project
