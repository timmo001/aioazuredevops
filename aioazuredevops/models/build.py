"""DevOps Builds.

https://docs.microsoft.com/en-gb/rest/api/azure/devops/build/builds?view=azure-devops-rest-7.2-preview-preview
"""

from dataclasses import dataclass

from .core import Project


@dataclass
class BuildLinks:
    """DevOps build links."""

    l_self: str | None = None
    web: str | None = None
    source_version_display_uri: str | None = None
    timeline: str | None = None
    badge: str | None = None


@dataclass
class BuildDefinition:
    """DevOps build definition."""

    build_id: int
    name: str
    url: str | None = None
    path: str | None = None
    build_type: str | None = None
    queue_status: str | None = None
    revision: int | None = None


@dataclass
class Build:
    """DevOps build."""

    build_id: int
    build_number: str | None = None
    status: str | None = None
    result: str | None = None
    source_branch: str | None = None
    source_version: str | None = None
    priority: str | None = None
    reason: str | None = None
    queue_time: str | None = None
    start_time: str | None = None
    finish_time: str | None = None
    definition: BuildDefinition | None = None
    project: Project | None = None
    links: BuildLinks | None = None
