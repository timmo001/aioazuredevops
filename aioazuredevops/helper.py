"""Helper functions for Azure DevOps."""

from dataclasses import dataclass

from .models.iteration import Iteration, IterationTimeFrame
from .models.work_item import WorkItem
from .models.work_item_type import Category, State, WorkItemType


def current_iteration(iterations: list[Iteration]) -> Iteration | None:
    """Get current iteration."""
    for iteration in iterations:
        if iteration.attributes.time_frame == IterationTimeFrame.CURRENT:
            return iteration
    return None


def previous_iteration(iterations: list[Iteration]) -> Iteration | None:
    """Get previous iteration."""
    for key, iteration in enumerate(iterations):
        if iteration.attributes.time_frame == IterationTimeFrame.CURRENT:
            return iterations[key - 1]
    return None


def next_iteration(iterations: list[Iteration]) -> Iteration | None:
    """Get next iteration."""
    for key, iteration in enumerate(iterations):
        if iteration.attributes.time_frame == IterationTimeFrame.CURRENT:
            return iterations[key + 1]
    return None


@dataclass
class WorkItemState(State):
    """Work item by type and state."""

    work_items: list[WorkItem]


@dataclass
class WorkItemTypeAndState(WorkItemType):
    """Work item by type and state."""

    state_items: list[WorkItemState]


def work_item_types_states_filter(
    work_item_types: list[WorkItemType],
    categories: list[Category] | None = None,
    ignored_categories: list[Category] | None = None,
) -> list[str]:
    """Get states filter by category."""
    if categories is None and ignored_categories is None:
        return []

    states = []
    for work_item_type in work_item_types:
        for state in work_item_type.states:
            if categories is not None and state.category in categories:
                states.append(state.name)
            elif (
                ignored_categories is not None
                and state.category not in ignored_categories
            ):
                states.append(state.name)
    return states


def work_items_by_type_and_state(
    work_item_types: list[WorkItemType],
    work_items: list[WorkItem],
    categories: list[Category] | None = None,
    ignored_categories: list[Category] | None = None,
) -> list[WorkItemTypeAndState]:
    """Get work items by type and state."""

    result: list[WorkItemTypeAndState] = []
    for work_item_type in work_item_types:
        states: list[WorkItemState] = []
        for state in work_item_type.states:
            items = [item for item in work_items if item.fields.state == state.name]
            state = WorkItemState(
                name=state.name,
                color=state.color,
                category=state.category,
                work_items=items,
            )
            if categories is not None and state.category in categories:
                states.append(state)
            elif (
                ignored_categories is not None
                and state.category not in ignored_categories
            ):
                states.append(state)
        result.append(
            WorkItemTypeAndState(
                name=work_item_type.name,
                reference_name=work_item_type.reference_name,
                description=work_item_type.description,
                color=work_item_type.color,
                icon=work_item_type.icon,
                is_disabled=work_item_type.is_disabled,
                xml_form=work_item_type.xml_form,
                fields=work_item_type.fields,
                field_instances=work_item_type.field_instances,
                transitions=work_item_type.transitions,
                states=work_item_type.states,
                url=work_item_type.url,
                state_items=states,
            )
        )

    return result
