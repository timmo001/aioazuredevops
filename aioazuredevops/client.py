"""Get data from the Azure DevOps API"""
from datetime import datetime

import aiohttp

from aioazuredevops.builds import DevOpsBuild, DevOpsBuildDefinition, DevOpsBuildLinks
from aioazuredevops.core import DevOpsLinks, DevOpsProject, DevOpsTeam
from aioazuredevops.wiql import DevOpsWiqlColumn, DevOpsWiqlResult, DevOpsWiqlWorkItem
from aioazuredevops.work_item import (
    DevOpsWorkItem,
    DevOpsWorkItemAvatar,
    DevOpsWorkItemLinks,
    DevOpsWorkItemUser,
    DevOpsWorkItemValue,
    DevOpsWorkItemValueFields,
)


class DevOpsClient:
    """Client for Azure DevOps."""

    def __init__(self) -> None:
        """Initilalize."""
        self._authorized = False
        self._pat = None

    @property
    def authorized(self):
        """Is the client authorized."""
        return self._authorized

    @property
    def pat(self):
        """Get the PAT."""
        return self._pat

    async def get(
        self,
        session: aiohttp.ClientSession,
        url: str,
    ) -> aiohttp.ClientResponse:
        """Runs a GET request and returns response"""
        if self._pat is None:
            return await session.get(url)
        return await session.get(
            url,
            headers={"Authorization": aiohttp.BasicAuth("", self._pat).encode()},
        )

    async def post(
        self,
        session: aiohttp.ClientSession,
        url: str,
        data: dict,
    ) -> aiohttp.ClientResponse:
        """Runs a POST request and returns response"""
        if self._pat is None:
            return await session.post(
                url,
                json=data,
            )
        return await session.post(
            url,
            json=data,
            headers={"Authorization": aiohttp.BasicAuth("", self._pat).encode()},
        )

    async def authorize(
        self,
        pat: str,
        organization: str,
    ) -> None:
        """Authenticate."""
        async with aiohttp.ClientSession() as session:
            self._pat = pat
            response: aiohttp.ClientResponse = await self.get(
                session, f"https://dev.azure.com/{organization}/_apis/projects"
            )
            if response.status == 200:
                self._authorized = True
            else:
                self._authorized = False

    async def get_project(
        self,
        organization: str,
        project: str,
    ) -> DevOpsProject:
        """Get Azure DevOps project."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/_apis/projects/{project}",
            )
            if response.status != 200:
                return None
            json = await response.json()
            if json is None:
                return None

            return DevOpsProject(
                json["id"],
                json["name"],
                json["description"] if "description" in json else None,
                json["url"] if "url" in json else None,
                json["state"] if "state" in json else None,
                json["revision"] if "revision" in json else None,
                json["visibility"] if "visibility" in json else None,
                datetime.strptime(json["lastUpdateTime"], "%Y-%m-%dT%H:%M:%S.%fZ")
                if "lastUpdateTime" in json
                else None,
                DevOpsTeam(
                    json["defaultTeam"]["id"],
                    json["defaultTeam"]["name"],
                    json["defaultTeam"]["url"],
                )
                if "defaultTeam" in json
                else None,
                DevOpsLinks(
                    json["_links"]["self"]["href"],
                    json["_links"]["collection"]["href"],
                    json["_links"]["web"]["href"],
                )
                if "_links" in json
                else None,
            )

    async def get_builds(
        self,
        organization: str,
        project: str,
        parameters: str,
    ) -> list[DevOpsBuild]:
        """Get Azure DevOps builds."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/build/builds{parameters}",
            )
            if response.status != 200:
                return None
            json = await response.json()
            if json is None:
                return None

            builds = []
            for build in json["value"]:
                builds.append(
                    DevOpsBuild(
                        build["id"],
                        build["buildNumber"] if "buildNumber" in build else None,
                        build["status"] if "status" in build else None,
                        build["result"] if "result" in build else None,
                        build["sourceBranch"] if "sourceBranch" in build else None,
                        build["sourceVersion"] if "sourceVersion" in build else None,
                        build["priority"] if "priority" in build else None,
                        build["reason"] if "reason" in build else None,
                        build["queueTime"] if "queueTime" in build else None,
                        build["startTime"] if "startTime" in build else None,
                        build["startTime"] if "startTime" in build else None,
                        DevOpsBuildDefinition(
                            build["definition"]["id"],
                            build["definition"]["name"],
                            build["definition"]["url"] if "url" in build["definition"] else None,
                            build["definition"]["path"] if "path" in build["definition"] else None,
                            build["definition"]["type"] if "type" in build["definition"] else None,
                            build["definition"]["queueStatus"]
                            if "queueStatus" in build["definition"]
                            else None,
                            build["definition"]["revision"]
                            if "revision" in build["definition"]
                            else None,
                        )
                        if "definition" in build
                        else None,
                        DevOpsProject(
                            build["project"]["id"],
                            build["project"]["name"],
                            build["project"]["description"]
                            if "description" in build["project"]
                            else None,
                            build["project"]["url"] if "url" in build["project"] else None,
                            build["project"]["state"] if "state" in build["project"] else None,
                            build["project"]["revision"]
                            if "revision" in build["project"]
                            else None,
                            build["project"]["visibility"]
                            if "visibility" in build["project"]
                            else None,
                            datetime.strptime(
                                build["project"]["lastUpdateTime"],
                                "%Y-%m-%dT%H:%M:%S.%fZ",
                            )
                            if "lastUpdateTime" in build["project"]
                            else None,
                        )
                        if "project" in build
                        else None,
                        DevOpsBuildLinks(
                            build["_links"]["self"]["href"] if "self" in build["_links"] else None,
                            build["_links"]["web"]["href"] if "web" in build["_links"] else None,
                            build["_links"]["sourceVersionDisplayUri"]["href"]
                            if "sourceVersionDisplayUri" in build["_links"]
                            else None,
                            build["_links"]["timeline"]["href"]
                            if "timeline" in build["_links"]
                            else None,
                            build["_links"]["badge"]["href"]
                            if "badge" in build["_links"]
                            else None,
                        )
                        if "_links" in build
                        else None,
                    )
                )

            return builds

    async def get_build(
        self,
        organization: str,
        project: str,
        build_id: int,
    ) -> DevOpsBuild:
        """Get Azure DevOps builds."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/build/builds/{build_id}",
            )
            if response.status != 200:
                return None
            build = await response.json()
            if build is None:
                return None

            return DevOpsBuild(
                build["id"],
                build["buildNumber"] if "buildNumber" in build else None,
                build["status"] if "status" in build else None,
                build["result"] if "result" in build else None,
                build["sourceBranch"] if "sourceBranch" in build else None,
                build["sourceVersion"] if "sourceVersion" in build else None,
                build["priority"] if "priority" in build else None,
                build["reason"] if "reason" in build else None,
                build["queueTime"] if "queueTime" in build else None,
                build["startTime"] if "startTime" in build else None,
                build["startTime"] if "startTime" in build else None,
                DevOpsBuildDefinition(
                    build["definition"]["id"],
                    build["definition"]["name"],
                    build["definition"]["url"] if "url" in build["definition"] else None,
                    build["definition"]["path"] if "path" in build["definition"] else None,
                    build["definition"]["type"] if "type" in build["definition"] else None,
                    build["definition"]["queueStatus"]
                    if "queueStatus" in build["definition"]
                    else None,
                    build["definition"]["revision"] if "revision" in build["definition"] else None,
                )
                if "definition" in build
                else None,
                DevOpsProject(
                    build["project"]["id"],
                    build["project"]["name"],
                    build["project"]["description"] if "description" in build["project"] else None,
                    build["project"]["url"] if "url" in build["project"] else None,
                    build["project"]["state"] if "state" in build["project"] else None,
                    build["project"]["revision"] if "revision" in build["project"] else None,
                    build["project"]["visibility"] if "visibility" in build["project"] else None,
                    datetime.strptime(
                        build["project"]["lastUpdateTime"],
                        "%Y-%m-%dT%H:%M:%S.%fZ",
                    )
                    if "lastUpdateTime" in build["project"]
                    else None,
                )
                if "project" in build
                else None,
                DevOpsBuildLinks(
                    build["_links"]["self"]["href"] if "self" in build["_links"] else None,
                    build["_links"]["web"]["href"] if "web" in build["_links"] else None,
                    build["_links"]["sourceVersionDisplayUri"]["href"]
                    if "sourceVersionDisplayUri" in build["_links"]
                    else None,
                    build["_links"]["timeline"]["href"] if "timeline" in build["_links"] else None,
                    build["_links"]["badge"]["href"] if "badge" in build["_links"] else None,
                )
                if "_links" in build
                else None,
            )

    async def get_work_items_ids_all(
        self,
        organization: str,
        project: str,
    ) -> DevOpsWiqlResult:
        """Get Azure DevOps work item ids from wiql."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.post(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/wit/wiql?api-version=6.0",
                {
                    "query": "Select [System.Id] From WorkItems",
                },
            )
            if response.status != 200:
                return None
            data = await response.json()
            if data is None:
                return None

            return DevOpsWiqlResult(
                query_type=data["queryType"],
                query_result_type=data["queryResultType"],
                as_of=data["asOf"],
                columns=[
                    DevOpsWiqlColumn(
                        reference_name=column["referenceName"],
                        name=column["name"],
                        url=column["url"],
                    )
                    for column in data["columns"]
                ],
                work_items=[
                    DevOpsWiqlWorkItem(
                        id=work_item["id"],
                        url=work_item["url"],
                    )
                    for work_item in data["workItems"]
                ],
            )

    async def get_work_items(
        self,
        organization: str,
        project: str,
        ids: list[int],
    ) -> DevOpsWorkItem:
        """Get Azure DevOps work items."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/wit/workitems?ids={','.join(str(id) for id in ids)}&api-version=6.0",
            )
            if response.status != 200:
                return None
            data = await response.json()
            if data is None:
                return None

            return DevOpsWorkItem(
                count=data["count"],
                value=[
                    DevOpsWorkItemValue(
                        id=work_item["id"],
                        rev=work_item["rev"],
                        fields=DevOpsWorkItemValueFields(
                            area_path=work_item["fields"]["System.AreaPath"],
                            team_project=work_item["fields"]["System.TeamProject"],
                            iteration_path=work_item["fields"]["System.IterationPath"],
                            work_item_type=work_item["fields"]["System.WorkItemType"],
                            state=work_item["fields"]["System.State"],
                            reason=work_item["fields"]["System.Reason"],
                            assigned_to=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.AssignedTo"][
                                    "displayName"
                                ],
                                url=work_item["fields"]["System.AssignedTo"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.AssignedTo"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.AssignedTo"]["id"],
                                unique_name=work_item["fields"]["System.AssignedTo"]["uniqueName"],
                                image_url=work_item["fields"]["System.AssignedTo"]["imageUrl"],
                                descriptor=work_item["fields"]["System.AssignedTo"]["descriptor"],
                            )
                            if "System.AssignedTo" in work_item["fields"]
                            and work_item["fields"]["System.AssignedTo"]
                            else None,
                            created_date=work_item["fields"]["System.CreatedDate"],
                            created_by=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.CreatedBy"]["displayName"],
                                url=work_item["fields"]["System.CreatedBy"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.CreatedBy"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.CreatedBy"]["id"],
                                unique_name=work_item["fields"]["System.CreatedBy"]["uniqueName"],
                                image_url=work_item["fields"]["System.CreatedBy"]["imageUrl"],
                                descriptor=work_item["fields"]["System.CreatedBy"]["descriptor"],
                            )
                            if "System.CreatedBy" in work_item["fields"]
                            and work_item["fields"]["System.CreatedBy"]
                            else None,
                            changed_date=work_item["fields"]["System.ChangedDate"],
                            changed_by=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.ChangedBy"]["displayName"],
                                url=work_item["fields"]["System.ChangedBy"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.ChangedBy"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.ChangedBy"]["id"],
                                unique_name=work_item["fields"]["System.ChangedBy"]["uniqueName"],
                                image_url=work_item["fields"]["System.ChangedBy"]["imageUrl"],
                                descriptor=work_item["fields"]["System.ChangedBy"]["descriptor"],
                            )
                            if "System.ChangedBy" in work_item["fields"]
                            and work_item["fields"]["System.ChangedBy"]
                            else None,
                            comment_count=work_item["fields"]["System.CommentCount"],
                            title=work_item["fields"]["System.Title"],
                            microsoft_vsts_common_state_change_date=work_item["fields"],
                            microsoft_vsts_common_priority=work_item["fields"][
                                "Microsoft.VSTS.Common.Priority"
                            ],
                        ),
                        url=work_item["url"],
                    )
                    for work_item in data["value"]
                ],
            )
