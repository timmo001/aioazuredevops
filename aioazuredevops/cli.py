"""Enable CLI"""
import asyncio

import click

from aioazuredevops.builds import DevOpsBuild
from aioazuredevops.client import DevOpsClient
from aioazuredevops.core import DevOpsProject
from aioazuredevops.wiql import DevOpsWiqlResult
from aioazuredevops.work_item import DevOpsWorkItem


@click.command()
@click.option("--organization", "-u", help="Organization")
@click.option("--project", "-p", help="Project")
@click.option("--pat", "-t", help="Personal Access Token")
def cli(organization: str, project: str, pat: str = None):
    """CLI for this package."""
    asyncio.get_event_loop().run_until_complete(handle(organization, project, pat))


async def handle(organization: str, project: str, pat: str = None) -> None:
    """Handle CLI command"""
    client = DevOpsClient()
    if pat is not None:
        await client.authorize(pat, organization)
        print(client.authorized)
        if client.authorized:
            print("Authorized.")
        else:
            print("Not authorized.")
            return
    print("Project:")
    doProject: DevOpsProject = await client.get_project(organization, project)
    if doProject is not None:
        print(doProject.id)
        print(doProject.name)
        print(doProject.description)
    print("Builds:")
    builds: list[DevOpsBuild] = await client.get_builds(
        organization,
        project,
        "?queryOrder=queueTimeDescending&maxBuildsPerDefinition=1",
    )
    if builds is not None and len(builds) > 0:
        build: DevOpsBuild = builds[0]
        if build is not None:
            print(build.id)
            print(build.build_number)
            print(build.status)
            print(build.result)
            print(build.source_branch)
            print(build.source_version)
            print(build.definition.id)
            print(build.definition.name)
            print(build.project.id)
            print(build.project.name)

    wiql_result: DevOpsWiqlResult = await client.get_work_items_ids_all(
        organization,
        project,
    )
    if wiql_result:
        ids: list[int] = [item.id for item in wiql_result.work_items]
        print("Work item ids:")
        print(ids)

        work_items: DevOpsWorkItem = await client.get_work_items(
            organization,
            project,
            ids,
        )
        if work_items:
            print("Work items:")
            for item in work_items.value:
                print("---------------------------------")
                print(item.id)
                print(item.fields.title)
                print(item.fields.created_by.display_name)
                print(item.fields.work_item_type)


cli()
