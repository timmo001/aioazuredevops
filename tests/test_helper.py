"""Test helpers."""

from aioresponses import aioresponses
import pytest
from syrupy.assertion import SnapshotAssertion

from aioazuredevops.client import DevOpsClient
from aioazuredevops.helper import (
    work_item_types_states_filter,
    work_items_by_type_and_state,
)
from aioazuredevops.models.work_item_types import Category

from . import ORGANIZATION, PROJECT


@pytest.mark.asyncio
async def test_work_item_types_states_filter(
    devops_client: DevOpsClient,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the get_work_item_types method."""
    work_item_types = await devops_client.get_work_item_types(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    assert work_item_types is not None

    assert work_item_types_states_filter(work_item_types) == snapshot(name="no_filter")

    assert (
        work_item_types_states_filter(
            work_item_types,
            categories=[Category.IN_PROGRESS],
            ignored_categories=[Category.COMPLETED, Category.REMOVED],
        )
        == snapshot()
    )


@pytest.mark.asyncio
async def test_work_items_by_type_and_state(
    devops_client: DevOpsClient,
    mock_aioresponse: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the get_work_item_types method."""
    work_item_types = await devops_client.get_work_item_types(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    assert work_item_types is not None

    work_item_ids = await devops_client.get_work_item_ids(
        organization=ORGANIZATION,
        project=PROJECT,
        states=work_item_types_states_filter(work_item_types),
    )

    assert work_item_ids is not None

    work_items = await devops_client.get_work_items(
        organization=ORGANIZATION,
        project=PROJECT,
        ids=work_item_ids,
    )

    assert work_items is not None

    assert (
        work_items_by_type_and_state(
            work_item_types,
            work_items,
            categories=[Category.IN_PROGRESS],
            ignored_categories=[Category.COMPLETED, Category.REMOVED],
        )
        == snapshot()
    )
