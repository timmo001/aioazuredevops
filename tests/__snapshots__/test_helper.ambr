# serializer version: 1
# name: test_work_item_types_states_filter.1
  list([
    'New',
    'On Hold',
    'Ready',
    'In Development',
    'In Code Review',
    'In Test',
    'In UAT',
    'In PreProd',
    'Ready for Release',
    'Released',
  ])
# ---
# name: test_work_item_types_states_filter[no_filter]
  list([
  ])
# ---
# name: test_work_items_by_type_and_state
  list([
    WorkItemTypeAndState(name='testname', reference_name='testreferenceName', description='testdescription', color='testcolor', icon=Icon(id='testid', url='testurl'), is_disabled=False, xml_form='testxmlForm', fields=[], field_instances=[], transitions={}, states=[State(name='New', color='b2b2b2', category=<Category.PROPOSED: 'Proposed'>), State(name='On Hold', color='fcfea8', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready', color='007acc', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Development', color='e87025', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Code Review', color='f9b978', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Test', color='fbd144', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In UAT', color='71338d', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In PreProd', color='ef33a3', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready for Release', color='207752', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Released', color='c3d84c', category=<Category.RESOLVED: 'Resolved'>), State(name='Closed', color='339933', category=<Category.COMPLETED: 'Completed'>)], url='testurl', state_items=[WorkItemState(name='New', color='b2b2b2', category=<Category.PROPOSED: 'Proposed'>, work_items=[]), WorkItemState(name='On Hold', color='fcfea8', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='Ready', color='007acc', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='In Development', color='e87025', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='In Code Review', color='f9b978', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='In Test', color='fbd144', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='In UAT', color='71338d', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='In PreProd', color='ef33a3', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='Ready for Release', color='207752', category=<Category.IN_PROGRESS: 'InProgress'>, work_items=[]), WorkItemState(name='Released', color='c3d84c', category=<Category.RESOLVED: 'Resolved'>, work_items=[])]),
  ])
# ---
